src/libraries/System.Net.Quic/tests/FunctionalTests/QuicListenerTests.cs

```diff
@@ -372,5 +372,67 @@ public async Task Listener_AwaitsConnection_ListenerSurvivesGC()
+                },
+                ConnectionOptionsCallback = (_, _, _) =>
+                {
+                    // Narrowing down alpn list to more specific.
+                    var options = CreateQuicServerOptions();
+                    options.ServerAuthenticationOptions.ApplicationProtocols = new()
+                    {
+                        new SslApplicationProtocol("bar"),
+                        new SslApplicationProtocol("test"),
+                    };
+                    return ValueTask.FromResult(options);
+                }
+            };
+            await using QuicListener listener = await CreateQuicListener(listenerOptions);
+
+            // Successful connection with bar ALPN
+            QuicClientConnectionOptions clientOptions1 = CreateQuicClientOptions(listener.LocalEndPoint);
+            clientOptions1.ClientAuthenticationOptions.ApplicationProtocols = new()
+            {
+                new SslApplicationProtocol("foo"),
+                new SslApplicationProtocol("bar"),
+            };
+            ValueTask<QuicConnection> connectTask1 = CreateQuicConnection(clientOptions1);
+            await using QuicConnection serverConnection1 = await listener.AcceptConnectionAsync().AsTask().WaitAsync(TimeSpan.FromSeconds(30));
+            await using QuicConnection clientConnection1 = await connectTask1.AsTask().WaitAsync(TimeSpan.FromSeconds(30));
+
+            Assert.Equal(new SslApplicationProtocol("bar"), clientConnection1.NegotiatedApplicationProtocol);
+
+            // Successful connection with test ALPN
+            QuicClientConnectionOptions clientOptions2 = CreateQuicClientOptions(listener.LocalEndPoint);
+            clientOptions2.ClientAuthenticationOptions.ApplicationProtocols = new()
+            {
+                new SslApplicationProtocol("foo"),
+                new SslApplicationProtocol("test"),
+            };
+            ValueTask<QuicConnection> connectTask2 = CreateQuicConnection(clientOptions2);
+            await using QuicConnection serverConnection2 = await listener.AcceptConnectionAsync().AsTask().WaitAsync(TimeSpan.FromSeconds(30));
+            await using QuicConnection clientConnection2 = await connectTask2.AsTask().WaitAsync(TimeSpan.FromSeconds(30));
+
+            Assert.Equal(new SslApplicationProtocol("test"), clientConnection2.NegotiatedApplicationProtocol);
+
+            // Failure case should throw QuicException
+            QuicClientConnectionOptions clientOptions3 = CreateQuicClientOptions(listener.LocalEndPoint);
+            clientOptions3.ClientAuthenticationOptions.ApplicationProtocols = new()
+            {
+                new SslApplicationProtocol("foo"),
+            };
+            ValueTask<QuicConnection> connectTask3 = CreateQuicConnection(clientOptions3);
+            await Assert.ThrowsAsync<QuicException>(() => listener.AcceptConnectionAsync().AsTask().WaitAsync(TimeSpan.FromSeconds(30)));
```

Comment:
I think we should be able to verify that QuicException is specifically for ALPN failure?

UPD: I also am a little bit confused now why it's QuicException, shouldn't ALPN failure be AuthenticationException? 🤔

https://github.com/dotnet/runtime/blob/da30389d2c0c3e5cad98f82e9bee0c9e8cfd78eb/src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/ThrowHelper.cs#L89

Reply:
That's what I'm investigating right now tbh, because for now, it returns:
```
System.Net.Quic.QuicException : An internal error has occurred. Status code: QUIC_STATUS_INTERNAL_ERROR
```
So looks like our status is something else than `QUIC_STATUS_ALPN_NEG_FAILURE`

Reply:
Cool, thanks for digging this up!

Reply:
throwing internal error on user data is not pleasant. Also we should verify what would happen if user selects something that was not in original list.